File: xdiff/xhistogram.c
/*
 * Copyright (C) 2010, Google Inc.
 * and other copyright owners as documented in JGit's IP log.
 *
 * This program and the accompanying materials are made available
 * under the terms of the Eclipse Distribution License v1.0 which
 * accompanies this distribution, is reproduced below, and is
 * available at http://www.eclipse.org/org/documents/edl-v10.php
 *
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or
 * without modification, are permitted provided that the following
 * conditions are met:
 *
 * - Redistributions of source code must retain the above copyright
 *   notice, this list of conditions and the following disclaimer.
 *
 * - Redistributions in binary form must reproduce the above
 *   copyright notice, this list of conditions and the following
 *   disclaimer in the documentation and/or other materials provided
 *   with the distribution.
 *
 * - Neither the name of the Eclipse Foundation, Inc. nor the
 *   names of its contributors may be used to endorse or promote
 *   products derived from this software without specific prior
 *   written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND
 * CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES,
 * INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
 * OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR
 * CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
 * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
 * NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
 * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
 * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
 * STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF
 * ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

#include "xinclude.h"

#define MAX_CHAIN_LENGTH 64

#define LINE_END(n) (line##n + count##n - 1)
#define ONE_INDEXED 1

struct record {
	size_t ptr, cnt;
	struct record *next;
};

DEFINE_IVEC_TYPE(struct record, record);
DEFINE_IVEC_TYPE(struct record*, record_ptr);

struct histindex {
	struct IVec_record_ptr record;   /* an occurrence */
	struct IVec_record_ptr line_map; /* map of line to record chain */
	struct IVec_record record_storage;
	struct IVec_usize next_ptr;
	size_t table_bits;
	size_t ptr_shift;
	size_t cnt;
	bool has_common;

	xdfenv_t *env;
};

struct region {
	size_t begin1, end1;
	size_t begin2, end2;
};

#define LINE_MAP(i, a) (i->line_map.ptr[(a) - i->ptr_shift])

#define NEXT_PTR(index, line_number) \
	(index->next_ptr.ptr[(line_number) - index->ptr_shift])

#define CNT(index, ptr) \
	((LINE_MAP(index, ptr))->cnt)

#define MPH(env, s, l) \
	(env->xdf##s.minimal_perfect_hash.ptr[l - ONE_INDEXED])

#define CMP(i, s1, l1, s2, l2) \
	(MPH(i->env, s1, l1) == MPH(i->env, s2, l2))

#define TABLE_HASH(index, side, line) \
	XDL_HASHLONG(MPH(index->env, side, line), index->table_bits)

static int scanA(struct histindex *index, size_t line1, size_t count1)
{
	size_t ptr, tbl_idx;
	size_t chain_len;
	struct record **rec_chain, *rec;

	for (ptr = LINE_END(1); line1 <= ptr; ptr--) {
		tbl_idx = TABLE_HASH(index, 1, ptr);
		rec_chain = &index->record.ptr[tbl_idx];
		rec = *rec_chain;

		chain_len = 0;
		while (rec) {
			if (CMP(index, 1, rec->ptr, 1, ptr)) {
				/*
				 * ptr is identical to another element. Insert
				 * it onto the front of the existing element
				 * chain.
				 */
				NEXT_PTR(index, ptr) = rec->ptr;
				rec->ptr = ptr;
				rec->cnt += 1;
				LINE_MAP(index, ptr) = rec;
				goto continue_scan;
			}

			rec = rec->next;
			chain_len++;
		}

		if (chain_len == MAX_CHAIN_LENGTH)
			return -1;

		/*
		 * This is the first time we have ever seen this particular
		 * element in the sequence. Construct a new chain for it.
		 */
		if (index->record_storage.capacity == 0)
			ivec_reserve_exact(&index->record_storage, index->env->mph_size);
		rec = &index->record_storage.ptr[index->record_storage.length++];
		rec->ptr = ptr;
		rec->cnt = 1;
		rec->next = *rec_chain;
		*rec_chain = rec;
		LINE_MAP(index, ptr) = rec;

continue_scan:
		; /* no op */
	}

	return 0;
}

static int try_lcs(struct histindex *index, struct region *lcs, size_t b_ptr,
	size_t line1, size_t count1, size_t line2, size_t count2)
{
	size_t b_next = b_ptr + 1;
	struct record *rec = index->record.ptr[TABLE_HASH(index, 2, b_ptr)];
	size_t as, ae, bs, be, np, rc;
	bool should_break;

	for (; rec; rec = rec->next) {
		if (rec->cnt > index->cnt) {
			if (!index->has_common)
				index->has_common = CMP(index, 1, rec->ptr, 2, b_ptr);
			continue;
		}

		as = rec->ptr;
		if (!CMP(index, 1, as, 2, b_ptr))
			continue;

		index->has_common = 1;
		for (;;) {
			should_break = 0;
			np = NEXT_PTR(index, as);
			bs = b_ptr;
			ae = as;
			be = bs;
			rc = rec->cnt;

			while (line1 < as && line2 < bs
				&& CMP(index, 1, as - 1, 2, bs - 1)) {
				as--;
				bs--;
				if (1 < rc)
					rc = XDL_MIN(rc, CNT(index, as));
			}
			while (ae < LINE_END(1) && be < LINE_END(2)
				&& CMP(index, 1, ae + 1, 2, be + 1)) {
				ae++;
				be++;
				if (1 < rc)
					rc = XDL_MIN(rc, CNT(index, ae));
			}

			if (b_next <= be)
				b_next = be + 1;
			if (lcs->end1 - lcs->begin1 < ae - as || rc < index->cnt) {
				lcs->begin1 = as;
				lcs->begin2 = bs;
				lcs->end1 = ae;
				lcs->end2 = be;
				index->cnt = rc;
			}

			if (np == 0)
				break;

			while (np <= ae) {
				np = NEXT_PTR(index, np);
				if (np == 0) {
					should_break = 1;
					break;
				}
			}

			if (should_break)
				break;

			as = np;
		}
	}
	return b_next;
}

static inline void free_index(struct histindex *index)
{
	ivec_free(&index->record);
	ivec_free(&index->line_map);
	ivec_free(&index->record_storage);
	ivec_free(&index->next_ptr);
}

static int histindex_init(struct histindex *index, xdfenv_t *env, size_t line1, size_t count1)
{
	memset(index, 0, sizeof(struct histindex));
	IVEC_INIT(index->record);
	IVEC_INIT(index->line_map);
	IVEC_INIT(index->record_storage);
	IVEC_INIT(index->next_ptr);

	index->env = env;

	index->table_bits = xdl_hashbits(count1);
	ivec_zero(&index->record, 1 << index->table_bits);

	ivec_zero(&index->line_map, count1);
	ivec_zero(&index->next_ptr, count1);

	index->ptr_shift = line1;

	return 0;
}

static int find_lcs(xdfenv_t *env,
		    struct region *lcs,
		    size_t line1, size_t count1, size_t line2, size_t count2)
{
	size_t b_ptr;
	int ret = -1;
	struct histindex index;

	histindex_init(&index, env, line1, count1);

	if (scanA(&index, line1, count1))
		goto cleanup;

	index.cnt = MAX_CHAIN_LENGTH + 1;

	for (b_ptr = line2; b_ptr <= LINE_END(2); )
		b_ptr = try_lcs(&index, lcs, b_ptr, line1, count1, line2, count2);

	if (index.has_common && MAX_CHAIN_LENGTH < index.cnt)
		ret = 1;
	else
		ret = 0;

cleanup:
	free_index(&index);
	return ret;
}

static int histogram_diff(uint64_t flags, xdfenv_t *env,
	size_t line1, size_t count1, size_t line2, size_t count2)
{
	struct region lcs;
	int lcs_found;
	int result;
redo:
	result = -1;

	if (count1 <= 0 && count2 <= 0)
		return 0;

	if (!count1) {
		while(count2--)
			env->xdf2.changed.ptr[line2++ - 1] = true;
		return 0;
	} else if (!count2) {
		while(count1--)
			env->xdf1.changed.ptr[line1++ - 1] = true;
		return 0;
	}

	memset(&lcs, 0, sizeof(lcs));
	lcs_found = find_lcs(env, &lcs, line1, count1, line2, count2);
	if (lcs_found < 0)
		goto out;
	else if (lcs_found)
		result = xdl_fall_back_diff(env, flags, line1, count1, line2, count2);
	else {
		if (lcs.begin1 == 0 && lcs.begin2 == 0) {
			while (count1--)
				env->xdf1.changed.ptr[line1++ - 1] = true;
			while (count2--)
				env->xdf2.changed.ptr[line2++ - 1] = true;
			result = 0;
		} else {
			result = histogram_diff(flags, env,
						line1, lcs.begin1 - line1,
						line2, lcs.begin2 - line2);
			if (result)
				goto out;
			/*
			 * result = histogram_diff(xpp, env,
			 *            lcs.end1 + 1, LINE_END(1) - lcs.end1,
			 *            lcs.end2 + 1, LINE_END(2) - lcs.end2);
			 * but let's optimize tail recursion ourself:
			*/
			count1 = LINE_END(1) - lcs.end1;
			line1 = lcs.end1 + 1;
			count2 = LINE_END(2) - lcs.end2;
			line2 = lcs.end2 + 1;
			goto redo;
		}
	}
out:
	return result;
}

int xdl_do_histogram_diff(xdfenv_t *env, uint64_t flags)
{
	size_t start = ONE_INDEXED + env->delta_start;
	size_t end1 = ONE_INDEXED + env->xdf1.record.length - env->delta_end;
	size_t end2 = ONE_INDEXED + env->xdf2.record.length - env->delta_end;

	return histogram_diff(flags, env,
		start, end1 - start,
		start, end2 - start);
}

File: rust/xdiff/src/xhistogram.rs
#![allow(non_camel_case_types)]

use std::marker::PhantomData;
use std::ops::Range;
use interop::ivec::IVec;
use crate::get_file_context;
use crate::xdiff::{LINE_SHIFT, SENTINEL, YES};
use crate::xdiffi::classic_diff_with_range;
use crate::xtypes::{xdpair, FileContext};


const MAX_CHAIN_LENGTH: usize = 64;


#[repr(C)]
struct record {
	line_number: usize,
    count: usize,
	next: *mut record,
}

impl Default for record {
	fn default() -> Self {
		Self {
			line_number: 0,
			count: 0,
			next: std::ptr::null_mut(),
		}
	}
}


struct RecordIter<'a> {
	cur: *mut record,
	_marker: PhantomData<&'a record>,
}

impl<'a> Iterator for RecordIter<'a> {
	type Item = &'a mut record;

	fn next(&mut self) -> Option<Self::Item> {
		if self.cur.is_null() {
			return None;
		}
		let t = self.cur;
		self.cur = unsafe { (*self.cur).next };
		Some(unsafe { &mut *t })
	}
}

impl<'a> RecordIter<'a> {
	fn new(start: *mut record) -> Self {
		Self {
			cur: start,
			_marker: PhantomData,
		}
	}
}


#[repr(C)]
struct histindex {
	record_storage: Vec<record>,
	record: IVec<*mut record>,
	line_map: IVec<*mut record>,
	next_line_numbers: IVec<usize>,
	line_number_shift: usize,
	count: usize,
	has_common: bool,
}


#[repr(C)]
struct region {
	range1: Range<usize>,
	range2: Range<usize>,
}


fn scan_a(index: &mut histindex, pair: &mut xdpair, range1: Range<usize>) -> i32 {
    let lhs = FileContext::new(&mut pair.lhs);

    'outer: for line_number in range1.rev() {
		let mph1 = lhs.minimal_perfect_hash[line_number - LINE_SHIFT] as usize;

		let mut chain_len = 0;
		for rec in RecordIter::new(index.record[mph1]) {
			let mph2 = lhs.minimal_perfect_hash[rec.line_number - LINE_SHIFT] as usize;
			if mph1 == mph2 {
				/*
				 * line_number is identical to another element. Insert
				 * it onto the front of the existing element
				 * chain.
				 */
				index.next_line_numbers[line_number - index.line_number_shift] = rec.line_number;
				rec.line_number = line_number;
				rec.count += 1;
				index.line_map[line_number - index.line_number_shift] = rec;
				continue 'outer;
			}

			chain_len += 1;
		}

		if chain_len == MAX_CHAIN_LENGTH {
			return -1;
		}

		/*
		 * This is the first time we have ever seen this particular
		 * element in the sequence. Construct a new chain for it.
		 */
		let last = index.record_storage.len();
		index.record_storage.push(record::default());
		let rec = &mut index.record_storage[last];
		rec.line_number = line_number;
		rec.count = 1;
		rec.next = index.record[mph1];
		index.record[mph1] = rec;
		index.line_map[line_number - index.line_number_shift] = rec;
	}

	0
}


fn try_lcs(index: &mut histindex, pair: &mut xdpair, lcs: &mut region, b_line_number: usize,
				  range1: Range<usize>, range2: Range<usize>,
) -> usize {
	let rhs = FileContext::new(&mut pair.rhs);

	let mut b_next = b_line_number + 1;
	let b_line_number_mph = rhs.minimal_perfect_hash[b_line_number - LINE_SHIFT] as usize;
	let mut range_a = Range::default();
	let mut range_b = Range::default();

	for rec in RecordIter::new(index.record[b_line_number_mph]) {
		if rec.count > index.count {
			if !index.has_common {
				index.has_common = pair.equal_by_line_number(rec.line_number, b_line_number);
			}
			continue;
		}

		range_a.start = rec.line_number;
		if !pair.equal_by_line_number(range_a.start, b_line_number) {
			continue;
		}

		index.has_common = true;
		'middle_loop: loop {
			let mut next_line = index.next_line_numbers[range_a.start - index.line_number_shift];
			range_b.start = b_line_number;
			range_a.end = range_a.start;
			range_b.end = range_b.start;
			let mut record_count = rec.count;

			while range1.start < range_a.start && range2.start < range_b.start
				&& pair.equal_by_line_number(range_a.start - 1, range_b.start - 1) {
				range_a.start -= 1;
				range_b.start -= 1;
				if 1 < record_count {
					let t_rec: *mut record = index.line_map[range_a.start - index.line_number_shift];
					let count = unsafe { (*t_rec).count };
					record_count = std::cmp::min(record_count, count);
				}
			}
			while range_a.end < range1.end - 1 && range_b.end < range2.end - 1
				&& pair.equal_by_line_number(range_a.end + 1, range_b.end + 1) {
				range_a.end += 1;
				range_b.end += 1;
				if 1 < record_count {
					let t_rec: *mut record = index.line_map[range_a.end - index.line_number_shift];
					let count = unsafe { (*t_rec).count };
					record_count = std::cmp::min(record_count, count);
				}
			}

			if b_next <= range_b.end {
				b_next = range_b.end + 1;
			}
			if lcs.range1.end - lcs.range1.start < range_a.end - range_a.start || record_count < index.count {
				lcs.range1.start = range_a.start;
				lcs.range2.start = range_b.start;
				lcs.range1.end = range_a.end;
				lcs.range2.end = range_b.end;
				index.count = record_count;
			}

			if next_line == 0 {
				break;
			}

			while next_line <= range_a.end {
				next_line = index.next_line_numbers[next_line - index.line_number_shift];
				if next_line == 0 {
					break 'middle_loop;
				}
			}

			range_a.start = next_line;
		}
	}

	b_next
}


fn find_lcs(pair: &mut xdpair, lcs: &mut region,
	range1: Range<usize>, range2: Range<usize>,
) -> i32 {
	let (lhs, rhs) = get_file_context!(pair);
	let fudge = (lhs.record.len() + rhs.record.len()) * 10;
	drop(lhs);
	drop(rhs);

	let mut index = histindex {
		record_storage: Vec::with_capacity(fudge),
		record: unsafe { IVec::zero(pair.minimal_perfect_hash_size) },
		line_map: unsafe { IVec::zero(range1.len()) },
		next_line_numbers: unsafe { IVec::zero(range1.len()) },
		line_number_shift: range1.start,
		count: 0,
		has_common: false,
	};

	if scan_a(&mut index, pair, range1.clone()) != 0 {
		return -1;
	}

	index.count = MAX_CHAIN_LENGTH + 1;

	let mut b_line_number = range2.start;
	while b_line_number < range2.end {
		b_line_number = try_lcs(&mut index, pair, lcs, b_line_number, range1.clone(), range2.clone());
	}

	if index.has_common && MAX_CHAIN_LENGTH < index.count {
		1
	} else {
		0
	}
}


fn histogram_diff(flags: u64, pair: &mut xdpair,
	mut range1: Range<usize>, mut range2: Range<usize>,
) -> i32 {
	let mut result;
	loop {
		result = -1;

		if range1.len() <= 0 && range2.len() <= 0 {
			return 0;
		}

		if range1.len() == 0 {
			for i in range2 {
				pair.rhs.consider[SENTINEL + i - LINE_SHIFT] = YES;
			}
			return 0;
		}
		if range2.len() == 0 {
			for i in range1 {
				pair.lhs.consider[SENTINEL + i - LINE_SHIFT] = YES;
			}
			return 0;
		}

		let mut lcs = region {
			range1: Range::default(),
			range2: Range::default(),
		};
		let lcs_found = find_lcs(pair, &mut lcs, range1.clone(), range2.clone());
		if lcs_found < 0 {
			return result;
		}

		if lcs_found != 0 {
			return classic_diff_with_range(flags, pair, range1, range2);
		}

		if lcs.range1.start == 0 && lcs.range2.start == 0 {
			for i in range1 {
				pair.lhs.consider[SENTINEL + i - 1] = YES;
			}
			for i in range2 {
				pair.rhs.consider[SENTINEL + i - 1] = YES;
			}
			result = 0;
		} else {
			result = histogram_diff(flags, pair,
						range1.start..lcs.range1.start,
						range2.start..lcs.range2.start);
			if result != 0 {
				return result;
			}
			/*
			 * result = histogram_diff(flags, pair,
			 *            lcs.end1 + 1, LINE_END(1) - lcs.end1,
			 *            lcs.end2 + 1, LINE_END(2) - lcs.end2);
			 * but let's optimize tail recursion ourself:
			*/
			range1.start = lcs.range1.end + 1;
			range2.start = lcs.range2.end + 1;
			continue;
		}
		break;
	}

	result
}


pub(crate) fn do_histogram_diff(flags: u64, pair: &mut xdpair) -> i32 {
	let mut range1 = Range::default();
	let mut range2 = Range::default();

	let (lhs, rhs) = get_file_context!(pair);
	range1.start = LINE_SHIFT + pair.delta_start;
	range1.end = LINE_SHIFT + lhs.record.len() - pair.delta_end;
	range2.start = LINE_SHIFT + pair.delta_start;
	range2.end = LINE_SHIFT + rhs.record.len() - pair.delta_end;

	histogram_diff(flags, pair, range1, range2)
}

